### PR TITLE
level 0 edtf refactoring

### PR DESCRIPTION
**Pull request by kilip**

### CODE REVIEW

File: tests/Unit/L1QualificationTest.php [hunk 1]
```diff
@@ -1,57 +0,0 @@
...
-use PHPUnit\Framework\TestCase;
-
-/**
- * Class L1DateTest
- * @covers \EDTF\Parser
- * @covers \EDTF\Interval
- * @covers \EDTF\ExtDateTime
- * @package EDTF\Tests\Unit
- */
-class L1QualificationTest extends TestCase
```
Comment (JeroenDeDauw) [inline tests/Unit/L1QualificationTest.php:16]: Why did you remove these tests?

File: src/EDTF.php [hunk 2]
```diff
@@ -4,13 +4,50 @@
 
 namespace EDTF;
 
-use EDTF\Contracts\DateTimeInterface;
-
 class EDTF
 {
-    public static function from(string $data): DateTimeInterface
+    public static function from(string $data): object
```
Comment (JeroenDeDauw) [inline src/EDTF.php:9]: Like this the caller still does not know how to handle the result

File: src/ExtDate.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+class ExtDate
+{
+    protected ?int $year;
+    protected ?int $month;
+    protected ?int $day;
+
+    private Parser $parser;
+
+    public function __construct(Parser $parser)
```
Comment (JeroenDeDauw) [inline src/ExtDate.php:16]: Like this you can't construct a `ExtDate` without having a `Parser`. That is really strange, especially after we talked about Value Objects. Why not this?

```php
__construct(?int $year, ?int $month, ?int $day)
```

Or preferably 


```php
__construct(int $year, int $month, int $day)
```

Comment (JeroenDeDauw) [reply src/ExtDate.php:16]: Optional fields are very suspect in objects like this. Especially all fields being optional.

File: src/ExtDateTime.php [hunk 4]
```diff
@@ -4,260 +4,67 @@
...
-                ->setTimezoneOffset($tzOffset);
+    private ?int $hour;
+    private ?int $minute;
+    private ?int $second;
+    private ?string $tzSign;
+    private ?int $tzMinute;
+    private ?int $tzHour;
+    private int $timezoneOffset = 0;
+
+    public function __construct(Parser $parser)
```
Comment (JeroenDeDauw) [inline src/ExtDateTime.php:17]: Same remark as for the other class - why is a Parser instance needed here?

File: src/Parser.php [hunk 5]
```diff
@@ -48,76 +46,100 @@ class Parser
 
 					# Others start #
 						(?>T # Literal T
-						(?<hourNum>2[0-3]|[01][0-9]):
-						(?<minuteNum>[0-5][0-9]):
-						(?<secondNum>[0-5][0-9])
+						(?<hour>2[0-3]|[01][0-9]):
+						(?<minute>[0-5][0-9]):
+						(?<second>[0-5][0-9])
```
Comment (JeroenDeDauw) [inline src/Parser.php:51]: This is the opposite of what you did in https://github.com/ProfessionalWiki/EDTF/commit/d3c0268be34466876480a04c6db1bb014acec7c3. Did you accidentally do a partial revert?

File: src/Parser.php [hunk 6]
```diff
@@ -48,76 +46,100 @@ class Parser
...
+    private ?int $year = null;
+    private ?int $month = null;
+    private ?int $day = null;
+    private ?int $hour = null;
+    private ?int $minute = null;
+    private ?int $second = null;
+    private ?string $tzSign = null;
+    private ?int $tzMinute = null;
+    private ?int $tzHour = null;
+    private ?string $tzUtc = null;
```
Comment (JeroenDeDauw) [inline src/Parser.php:69]: Adding these fields and corresponding getters here has me just as confused as requiring a Parser instance in the date obejcts. Why end up with such state in the Parser?

**Review by JeroenDeDauw — CHANGES_REQUESTED** [review]

File: tests/Unit/L1QualificationTest.php (hunk 1 above)
Comment (kilip) [reply tests/Unit/L1QualificationTest.php:16]: I just want to focus on Level 0 first. I still have a copy of this tests.

File: src/EDTF.php (hunk 2 above)
Comment (kilip) [reply src/EDTF.php:9]: This method can returns different object like an interval, season, set, etc. Do you have idea how to improve this method?

File: src/Parser.php (hunk 5 above)
Comment (kilip) [reply src/Parser.php:51]: No, I change this value so it can be easier to use in this [loop](https://github.com/ProfessionalWiki/EDTF/pull/5/files#diff-47bc8653b21716576958e25b6d7356ecb0f0070f17554d12d2bee985ac211b26R83-R91)

File: src/Parser.php (hunk 6 above)
Comment (kilip) [reply src/Parser.php:69]: I want a ```Parser``` just to do the parsing and store the results. This results should be then used by EDTF class to generate an ```ExtDate```, ```ExtDateTime```, or an ```Interval``` object

Comment (JeroenDeDauw) [reply src/Parser.php:69]: I think it would be better if this state was put in a dedicated object. That way the parser could hand that object to the EDTF class and remain state free itself. Having such state in a service like Parser is a design anti-pattern IMO. It opens the door for a whole class of state bugs and makes construction of the EDTF objects unnecessarily difficult.

At the same time I'd be weird for the Parser to return an intermediate object. Why can't it return the real object model (`ExtDate`, `ExtDateTime`, `Interval`) right away?

File: src/EDTF.php (hunk 2 above)
Comment (JeroenDeDauw) [reply src/EDTF.php:9]: Yeah good question. I do not have a good solution readily at hand.

The use case to cater to here is that the caller of the parser (or this static function) should know somehow which methods it can call. One approach would be `instanceof` checks. Another is to have methods like `isInternal(): bool {}` combined with `getInterval(): Interval {}`. 

If you can't think of anything better than `instanceof` checks we could just do that for now. Would still want to return an actual interface here though, so methods common to all EDTF classes can still be called without needing to do `instanceof`. That would be in line with edtf.js, where all these types implement https://github.com/inukshuk/edtf.js/blob/master/src/interface.js.

File: src/Parser.php (hunk 6 above)
Comment (kilip) [reply src/Parser.php:69]: Actually the old Parser returns a real object (```ExtDate```, ```ExtDateTime```, ```Interval```) like you mention above:

https://github.com/ProfessionalWiki/EDTF/blob/d3c0268be34466876480a04c6db1bb014acec7c3/src/Parser.php#L61-L67

I have change it, because you suggest ```Parser::parse()``` to returns a ```ParserResult``` object like in this pull:
https://github.com/ProfessionalWiki/EDTF/pull/4
Instead creating new Class I am thinking why not just having this Parser class to store all of the parsing result values. That's why this Parser class have properties like that.

File: src/EDTF.php (hunk 2 above)
Comment (kilip) [reply src/EDTF.php]: I think we can just use an Interface. I have added a new ```ExtDateInterface``` for this.

File: src/Parser.php (hunk 6 above)
Comment (JeroenDeDauw) [reply src/Parser.php:70]: Normally classes should only do one thing. Or have a **S**ingle Responsibility, the S in SOLID. Don't be afraid to create many classes or small classes, this is nearly always better than mashing things together or otherwise going against SOLID and similar standard software design practices. (Most devs don't know "standard" about design principles so the "standard" perhaps needs some quotes.)

I think the Parser should return an EDTF date, which is a Value Object. Possibly the date can be wrapped in a result object like the one I created an example off. The Parser should not expose internal state or require the caller to use another service such as a factory to turn the return value into an EDTF date. The parser _can_ be split up into multiple classes internally (if this makes things easier), but this should not be visible to the caller of the Parser (including most of the tests).

I have been wondering for some time what the `EDTF` class is all about. It does not seem needed. On your branch at the moment my impression is that it violates Single Responsibility by splitting up a responsibility over multiple classes. Classes full of static code are also very suspect. OO is about objects, not classes :)

Comment (kilip) [reply src/Parser.php:70]: So your decision is to let Parser act as factory like in this code?
https://github.com/ProfessionalWiki/EDTF/blob/d3c0268be34466876480a04c6db1bb014acec7c3/src/Parser.php#L61-L67

Comment (JeroenDeDauw) [reply src/Parser.php:70]: Yes

Comment (kilip) [reply src/Parser.php:72]: Done in https://github.com/ProfessionalWiki/EDTF/pull/5/commits/145d95a4d9092dbd9e10af2ff4d8fe8077eb33c7
